text/3605-freeze-uninit.md

```diff
@@ -0,0 +1,219 @@
+    * Undefined behaviour does not prevent malicious code from accessing any memory it physically can.
+    * That said, making such code UB could still be useful as it makes it unambiguously a bug to expose the contents of uninitialized memory to safe code, which can avoid accidental information leaks. If this RFC gets accepted, we should find ways to make it clear that even if doing so would be technically *sound*, this is still not something that Rust libraries are "supposed" to do and it must always be explicitly called out in the documentation.
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* The intrinsic could be defined as a by-value operation used by `MaybeUninit::freeze`, instead of the read operation used by `core::ptr::read_freeze`
+    * As noted above, it is believed that the two intrinsic choices are functionally identical, and thus the choice between them is completely arbitrary.
+* Either one of the two functions could be provided on their own
+    * Both functions are provided for maximum flexibility, and can be defined in terms of each other. The author does not believe there is significant drawback to providing both functions instead of just one
+* An in-place, mutable `freeze` could be offered, e.g. `MaybeUninit::freeze(&mut self)`
+    * While this function would seem to be a simple body that llvm could replace with a runtime no-op, in reality it is possible for [virtual memory](https://man7.org/linux/man-pages/man2/madvise.2.html#MADV_FREE) that has been freshly allocated and has not been written to exhibit properties of uninitialized memory (rather than simply being an abstract phenomenon the compiler tracks that disappears at runtime). Thus, such an operation would require a potentially expensive in-place copy. Until such time as an optimized version is available, we should avoid defining the in-place version, and require users to spell it explicitly as `*self = core::mem::replace(&mut self, uninit()).freeze()`.
+    * If intermediate representations are cooperative, it may be beneficial to provide the operation in the future, as it could perform only the writes required to ensure the backing memory is in a stable state (such as 1 write every 4096 bytes)
+    * Note that while `MaybeUninit::freeze(&mut self)` is possible to write, there is no `MaybeUninit::freeze(&self)`. This is because freezing uninit bytes requires performing writes in the abstract machine, overwriting initialized bytes with uninitialized ones, which are incompatible with the immutable `&Self` reciever.
```

Comment:
```suggestion
    * Note that while `MaybeUninit::freeze(&mut self)` is possible to write, there is no `MaybeUninit::freeze(&self)`. This is because freezing uninit bytes requires performing writes in the abstract machine, overwriting uninitialized bytes with initialized ones, which are incompatible with the immutable `&Self` reciever.
```